### PR TITLE
5 Sphinx integration

### PR DESCRIPTION
**Pull request by CKevin9**
- Install [Sphinx](https://www.sphinx-doc.org/en/master/usage/installation.html) into virtual environment
- Generate class and function description in documentation

### CODE REVIEW

File: doc/source/conf.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+# For the full list of built-in configuration values, see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+# -- Project information -----------------------------------------------------
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information
+import sys
+sys.path.insert(0, '/home/k377chen/Projects/PyVCAM/src/pyvcam/')
+
+project = 'PyVCAM'
+copyright = '2023, QuantumIon'
```
Comment (collineps) [inline doc/source/conf.py:12]: I don't think we actually have a copyright? Please omit this for now. Need to follow up with Crystal and Rajibul on our IP policy.

**Review by collineps — CHANGES_REQUESTED** [review]
This is excellent work! Thanks for leading the way on this. Please make the changes indicated in the review, as well as:

- [ ] Add a section in the README with instructions on how to build the documentation.
- [ ] While you're in the README, please update the `get_exp_modes` example to use the string key (we support that now, right?). Looks like we missed that in the last PR.

Also, going forward, please put a description of the changes made in the PR as the initial comment when you open the PR.

File: doc/source/conf.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Configuration file for the Sphinx documentation builder.
+#
+# For the full list of built-in configuration values, see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+# -- Project information -----------------------------------------------------
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information
+import sys
+sys.path.insert(0, '/home/k377chen/Projects/PyVCAM/src/pyvcam/')
```
[the remark this answers is not in the document]
Comment (CKevin9) [reply doc/source/conf.py:9]: Removed import statement - it is unnecessary in our code as we have made PyVCAM an installable

File: doc/source/usage.rst [hunk 3]
```diff
@@ -0,0 +1,30 @@
+Usage
+=====
+
+Saving an Image to Disk
+-----------------------
+
+To save an image to disk, the ``Image`` module from the Python library ``PIL`` has to be imported first.
```
[the remark this answers is not in the document]
Comment (CKevin9) [reply doc/source/usage.rst:7]: Yes, `PIL` is replaced by `Pillow`, but the import statement remains the same as shown in the [Pillow documentation](https://pillow.readthedocs.io/en/stable/handbook/tutorial.html)

[the remark this answers is not in the document]
Comment (collineps) [reply doc/source/usage.rst:7]: This is true. Just be explicit about which library you're using.

File: README.md [hunk 4]
```diff
@@ -87,9 +87,19 @@ This is an example of how to change some of the settings on the cameras.
 
 | Bare PyVCAM Class                   | Driver Wrapper                         |
 |-------------------------------------|----------------------------------------|
-| `cam.exp_mode = "Internal Trigger"` | `self.pyvcam.set_exp_mode(1792)`       |
+| `cam.exp_mode = "Internal Trigger"` | `self.pyvcam.set_exp_mode(1792)` or<br/>`self.pyvcam.set_exp_mode("Internal Trigger")`|
```
Comment (collineps) [inline README.md:90]:
```suggestion
| `cam.exp_mode = "Internal Trigger"` | `self.pyvcam.set_exp_mode("Internal Trigger")`|
```
Let's just go ahead and prefer the more explicit option.

File: docs/Makefile [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Minimal makefile for Sphinx documentation
+#
```
Comment (collineps) [inline docs/Makefile:2]:
```suggestion
```
Who needs it.

File: docs/make.bat
Comment (collineps) [inline docs/make.bat:1]: Is the rest of the package compatible with Windows as well?

File: docs/source/conf.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+#
+# For the full list of built-in configuration values, see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+# -- Project information -----------------------------------------------------
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information
+
+project = 'PyVCAM'
+author = 'Teledyne Photometrics, QuantumIon'
+release = '0.1'
```
Comment (collineps) [inline docs/source/conf.py:11]: Hm, the question of versioning is a pretty good one... We'll come back to that.

File: docs/source/conf.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+# -- General configuration ---------------------------------------------------
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration
+
+extensions = ['sphinx.ext.autodoc']
+
+templates_path = ['_templates']
+exclude_patterns = []
+
+
```
Comment (collineps) [inline docs/source/conf.py:21]:
```suggestion
```
Formatting.

File: docs/source/usage.rst [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+Saving an Image to Disk
+-----------------------
+
+To capture an image, the :meth:`get_frame() <pyvcam.driver.PyVCAM.get_frame>` function returns a 2D Numpy array.
+To convert the returned array to a visible image, additional tools are required, one of which is shown as an example below.
+
+Using the Pillow library
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+To save an image to disk, the ``Image`` class from the Python library ``PIL`` has to be imported first.
```
Comment (collineps) [inline docs/source/usage.rst:13]:
```suggestion
One method of saving an image to disk uses the ``Image`` class from ``[pillow](https://python-pillow.org/)``.
```
Only one method of many possible.

File: src/pyvcam/driver.py [hunk 9]
```diff
@@ -103,200 +110,221 @@ def finish(self) -> None:
...
-        Changes exposure resolution.
+        Changes exposure resolution. Refer to :func:`exp_resolutions` for the available exposure resolutions.
+
+        Default exposure resolutions:
+
         Keys:
-            0 - One Millisecond
-            1 - One Microsecond
+            | 0 - One Millisecond
+            | 1 - One Microsecond
```
Comment (collineps) [inline src/pyvcam/driver.py:229]:
```suggestion
            | 0 - One millisecond
            | 1 - One microsecond
```
Formatting.

**Review by collineps — CHANGES_REQUESTED** [review]
Great job. Please make the few frankly pedantic changes requested and then we're good to go!

File: src/pyvcam/driver.py (hunk 9 above)
Comment (CKevin9) [reply src/pyvcam/driver.py:229]: If the user wishes to change this setting by key strings, they will have to enter the key exactly, including the uppercase M's.
Of course, we could use Python's `.title()` function to ignore upper/lowercase user input, but the problem is replicated in the `set_exp_mode()` function where `Internal Trigger` and `Edge Trigger` keys are in title case, but `Trigger first` is only capitalized. These values are intrinsic to the `camera.py` file which reads from the `pvc` wrapper and cannot be changed (which is super annoying).

Comment (collineps) [reply src/pyvcam/driver.py:229]: Ah, I misunderstood! Thank you for the explanation. In that case, of course leave as is. Maybe add a comment here and in `set_exp_mode` that exact case matching is required.

If we want to really take care of the user, we could modify the input as you describe... Might be overkill.

**Pull request merged**